tools/autograd/templates/python_torch_functions.cpp

```diff
@@ -465,6 +467,85 @@ static PyObject * THPVariable_get_device(PyObject* self_, PyObject* args, PyObje
   END_HANDLE_TH_ERRORS
 }
 
+static PyObject * THPVariable_frombuffer(PyObject* self_, PyObject* args, PyObject* kwargs)
+{
+  HANDLE_TH_ERRORS
+  static PythonArgParser parser({
+    "frombuffer(PyObject* buffer, ScalarType dtype=None, int64_t count=-1, int64_t offset=0, *, bool requires_grad=False)",
+  }, /*traceable=*/false);
+
+  PyObject* ret = nullptr;
+  ParsedArgs<5> parsed_args;
+  auto r = parser.parse(args, kwargs, parsed_args);
+
+  if (r.idx == 0) {
+    auto buffer = r.pyobject(0);
+    auto dtype = r.scalartype(1);
+    auto count = r.toInt64(2);
+    auto offset = r.toInt64(3);
+    auto requires_grad = r.toBool(4);
+
+    auto elsize = at::elementSize(dtype);
+    size_t actual_count = 0;
+    Py_buffer view;
+
+    if (PyObject_GetBuffer(buffer, &view, PyBUF_SIMPLE) < 0) {
```

Comment:
Apparently, I missed the fact that buffers can be read-only. So, I probably should add a `PyBUF_WRITABLE` to ask for a writable there. That said, what should we do when the buffer is not writable? Should we just raise an error? Can tensors be read-only?

As is, it's possible to change read-only buffers. NumPy correctly raises an error when trying to modify the read-only buffer:

```python
> x = b"\x01\x02"
> a = torch.frombuffer(x, dtype=torch.uint8)
> a[0] = 10
> a
b'\n\x01'
```

```python
> a = numpy.frombuffer(x, dtype=torch.uint8)
> a[0] = 10
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: assignment destination is read-only
```

Reply:
PyTorch doesn't support read-only tensors. What about being consistent with the behavior of `from_numpy()` for non-writable arrays?

```
a = np.array((1., 2., 3.))
a.flags['WRITEABLE'] = False
torch.from_numpy(a)
: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:180.)
```

So similarly we could warn the buffer is not writeable, and PyTorch does not support non-writeable tensors?